### ISSUE TITLE
Sqawk

### ISSUE THREAD

**Issue by dbohdan** [thechangelog/ping#132]
**Sqawk** is an Awk-like command-line tool that makes it easy to use SQL on plain text data files. With it you can say things like

    find . -iname '*.jpg' -type f -printf '%s\n' | sqawk 'select sum(a1)/1024/1024 from a'

to sum up file sizes or

    sqawk -1 'select a0, "repeats", count(*), "times" from a group by a0 having count(*) > 1' file

to count duplicate lines in a file or

    sqawk 'select a1, b1, a2 from a inner join b on a2 = b2 where b1 < 10000 order by b1' MD5SUMS du-bytes

to combine data from two files, select those items where a particular field from the second file has a value less than 10000 then output the combined data sorted by that field. Sqawk is powered by SQLite.

https://github.com/dbohdan/sqawk

**Comment by jerodsanto** [comment]
@dbohdan cool stuff! How close is this to #93?

**Comment by dbohdan** [comment]
Pretty close! From reading the source code and the documentation for termsql here are the main advantages that I think the two project have over each other:

## Sqawk
* Can perform table joins on multiple files;
* Can split the incoming data into records and fields using a regex rather than a simple delimiter (termsql has literal delimiters and `--line-as-column` as a special case);
* Always keeps the original line in the column `a0` (like `$0` in Awk). This way  you can output raw unsplit text as you got it (preserving, e.g, the indentation) based on some conditionals on the split fields.

## termsql
* Can use named columns for more readable queries;
* Has a number of convenience command line options such as ones for skipping N lines at the begging or at the end of the input file, merging the last M columns into one (I like this!) and others;
* Supports CSV, TSV, HTML and SQL output.

I have also since found https://github.com/tjunier/sqawk (has table joins as well but works exclusively on CSV data) and https://github.com/samuel/squawk (a somewhat different concept: generates Python code for parsing structured data from SQL queries).

**Comment by jerodsanto** [comment]
@dbohdan awesome research. Gosh, naming things is hard. Come up with an awesome name like Sqawk and it's bound to have been used before.

**Comment by dbohdan** [comment]
>awesome research

Thanks!

>naming things is hard

Yep. I did my initial research on the name "Sqawk" with Google and only found https://www.openhub.net/p/sqawk, which is in a different niche. What's interesting is that apparently this exists: http://ivantomic.com/projects/ospnc/ — a tool for checking if a name for a FOSS project is already used by someone. I think your readers may appreciate this kind of thing (I would), so I made an issue for it: https://github.com/thechangelog/ping/issues/137.

**Comment by dbohdan** [comment]
Since the idea seems to be in the air maybe you could do an overview of command line tools for data processing? You can use the content of my comment above if you ever decide to.

**Comment by jerodsanto** [comment]
@dbohdan good idea. Interested in guest writing it for us?  We'll send you a :shirt:!

**Comment by dbohdan** [comment]
@jerodsanto I'm interested (I've had an article of that type on my mind lately) but I can't do it right now. What's the best way to contact you when I'm able to?

**Comment by jerodsanto** [comment]
@dbohdan excellent. Just head over to our [draft repo](https://github.com/thechangelog/draft) and follow the instructions in the README.

**Comment by dbohdan** [comment]
@jerodsanto Got it.

**Comment by jerodsanto** [comment]
@dbohdan I'm still keen on that blog post if you are, but for now we're going to link you up in our next issue of [Weekly](http://thechangelog.com/weekly/). It ships tomorrow, so subscribe if you haven't and keep your eyes peeled! :beers:

**Comment by danmbox** [comment]
@dbohdan Wouldn't it make sense to adopt the missing features from termsql (named columns, skipping lines, CSV output)?

**Comment by dbohdan** [comment]
@danmbox In light of your pull request and the recent influx of stargazers I implemented header rows and several other options in Sqawk that can be applied on a per-file basis. I've also refactored the code in a way that makes adding new features considerably easier.

I might implement CSV I/O (*I* more likely than *O*) later. As for skipping lines, I think that it is better left to `tail -n +$nlines`.
